tests/functional/u/use/used_before_assignment_issue2615.py

```diff
@@ -4,6 +4,57 @@ def main():
     try:
         res = 1 / 0
         res = 42
+        if main():
+            res = None
+        with open(__file__, encoding="utf-8") as opened_file:
+            res = opened_file.readlines()
     except ZeroDivisionError:
         print(res)  # [used-before-assignment]
     print(res)
+
+
+def nested_except_blocks():
+    """Assignments in an except are tested against possibly failing
+    assignments in try blocks at two different nesting levels."""
+    try:
+        res = 1 / 0
+        res = 42
+        if main():
+            res = None
+        with open(__file__, encoding="utf-8") as opened_file:
+            res = opened_file.readlines()
+    except ZeroDivisionError:
+        try:
+            more_bad_division = 1 / 0
+        except ZeroDivisionError:
+            print(more_bad_division)  # [used-before-assignment]
+            print(res)  # [used-before-assignment]
+    print(res)
+
+
+def consecutive_except_blocks():
+    """An assignment assumed to execute in one TryExcept should continue to be
+    assumed to execute in a consecutive TryExcept.
+    """
+    try:
+        res = 100
+    except ZeroDivisionError:
+        pass
+    try:
+        pass
+    except ValueError:
+        print(res)
```

Comment:
Same here, we want to avoid annoying users, and so outside a TryExcept we assume the try statements executed. By that same logic we assume that outside a TryExcept its except statements did not execute -- because it would be a bit backwards to assume the try statements never execute and thus that assignments in the except always happen.

Reply:
Didn't we say that like Pylance we consider that everything in a try/except can fail ? (https://github.com/PyCQA/pylint/pull/5384). Also, if we do not raise or return in the ``except`` it means we can reach line 43 from one branch or another and we would need to consider both branches in order to be correct.

Reply:
No, that wasn't my understanding, and that's not what Pylance does. The yellow underline is Pylance, showing that the checker we just worked on is in agreement with Pylance:
<img width="176" alt="Screen Shot 2021-12-23 at 10 44 31 AM" src="https://user-images.githubusercontent.com/38668450/147262922-2ecc043d-820d-4b4a-960f-2ea63dca3fc2.png">

Reply:
I guess a good way of putting it is that's why there are different private methods in this checker now:

`_uncertain_nodes_in_except_blocks`
`_uncertain_nodes_in_try_blocks_when_evaluating_except_blocks`
`_uncertain_nodes_in_try_blocks_when_evaluating_finally_blocks`

We make different control flow inferences based on where the name is being used (try, except, finally, or outside).

Reply:
Otherwise, we could just take everything in a try or except and make it uncertain. That would lead to lots of complaints :D

Reply:
I shouldn't make it sound more complicated than it really is 😃  -- before 2.13, Pylint assumed everything executed. Now, Pylint assumes the most likely cases execute: outside a try/except, assume the try block executed and the except did not, except when you have explicitly guarded against failure inside except/finally blocks, in which case you should assume the opposite.

Reply:
> Otherwise, we could just take everything in a try or except and make it uncertain. That would lead to lots of complaints :D

Why would that lead to a lot of complaints ? I think it's quite the contrary, pylint is supposed to detect potential problem like this. In the example you gave print(a) can fail if there was a ``ValueError``, right ? (The fact that ``a = 1`` can never raise a ValueError is irrelevant because we can't analyses code to tell if an Error can be raised, that would be a lot of effort). So  the *correct* thing to do is to consider that a ``ValueError`` can be raised - clearly the user think it can be raised, or they would not have added a try except. So ``a`` can be used before assignment. Maybe for such code the fix is to actually remove the try/except entirely, but clearly the code is wrong. Thoughts @DanielNoord ?

Reply:
I think assuming `try's` failed after the `except` is non-intuitive. For example:
```python
def func():
    """Returns bitwise value based on succes of operation"""
    try:
        return_value = make_request_to_server()
    except ServerError:
        print("Something failed")
        print("I will handle this gracefully")
        return 1
    return_value.handle_further()
    do_something_extra_with(return_value)
    return 2
```
I have chosen the names to be as expressive as possible.
It would be possible to rewrite this example to:
```python
def func():
    """Returns bitwise value based on succes of operation"""
    try:
        return_value = make_request_to_server()
        return_value.handle_further()
        do_something_extra_with(return_value)
        return 2
    except ServerError:
        print("Something failed")
        print("I will handle this gracefully")
        return 1
```
But that doesn't look very Pythonic to me. `try...except` are generally only around the functions that can actually fail instead of any functions that assume the `try` passed. I we were to assume the `try` failed you could only pass this check by writing example 2 (I think). That seems wrong to me.

Reply:
In your example you actually return in the `except` clause with `return 1` so if the `try` fail we never reach `return_value.handle_further()` and it's ok. What would you think of:
```python
def func():
    """Returns bitwise value based on succes of operation"""
    try:
        return_value = make_request_to_server()
    except ServerError:
        print("Something failed")
        print("I will handle this gracefully")
    return_value.handle_further() # [used-before-assignment] (if there is a ServerError)
```

Reply:
Ah I forgot about that.. That's a good point actually. Perhaps we should assume the `try` fails. I wonder why `Pylance` doesn't though. They probably have good reasons for this as well?
Perhaps we should do the partial assumption in `2.13`, wait for feedback and do the full assumption in `2.14`? That way we can see if we forgot anything based on community feedback.

Reply:
> In your example you actually return in the except clause with return 1 so if the try fail we never reach return_value.handle_further() and it's ok.

I think this illuminates the original problem. If we want to scale up what we're doing with this checker even further by assuming try statements are uncertain in every context, then we need a robust analysis of `raise` , `return` or `sys.exit` in each exception handler, as well as check name definitions in all except handlers. Because this example is fine:

```python
try:
    a = magic()
except ImportError:
    sys.exit(1)
except NameError:
    a = other_magic()
except ZeroDivisionError:
    print('zero!')
    raise
except Exception:
    a = constant
print(a)
```

That gets more tricky with things like this:

```python
try:
    a = magic()
except ZeroDivisionError:
    other_magic_that_might_return_or_raise_value_error()
except ValueError:
    a = None
print(a) # succeeds if ValueError was raised, otherwise crashes
```

The difficulty of these examples is what I was imagining when I was imagining complaints.

> Perhaps we should do the partial assumption in 2.13, wait for feedback and do the full assumption in 2.14?

Since we're talking about addressing false negatives, not false positives, that seems like a smart way of iterating to me.

***
Also, the test case that spawned this discussion is just adding extra coverage to make sure it didn't change -- there is no behavior change to that test case by this PR.

Reply:
Another way of looking at it: I think the issue we are discussing is essentially the same as #5524. Whether the try statement succeeds or fails is kind of irrelevant -- we should only raise an error after the except handlers if the except handlers do not each return, raise, or define the name.

Reply:
I agree with the points you're making here. Especially the last example is something of which I can easily imagine complaints coming from. Code with requests and databases seem to often really on such patterns.

Perhaps you could add a summary of this explanation to the issue you mentioned? That way we don't lose this.

Reply:
> Especially the last example is something of which I can easily imagine complaints coming from.

My last example was not quite right 😬 -- I neglected that subsequent handlers can't swallow exceptions raised in earlier handlers, so `ValueError` can't be passed down to a subsequent handler. I'll edit it.

Reply:
> Perhaps you could add a summary of this explanation to the issue you mentioned? That way we don't lose this.

I think this discussion is getting at when an `else` clause should be used in a try/except/else. If you can depend on exhaustive handling by your except handlers, you don't need to guard your `print(a)` after the try/except. If it's _not_ exhaustively handled, then your `print(a)` should be in an `else`.

I guess this could even be its own checker, so it could be enabled/disabled separately. (It would probably be easier to implement that way, given how much new logic would go into the exhaustiveness checking.)

Reply:
I think relying on everyone knowing that `library_function_that_might_sysexit` will implicitly exit all the time is an anti-pattern. Personally I would initialize the variable to be sure, or at least to add a comment stating explicitly that there's an exit and it's ok to do that if there's a good reason to do that like performance. (a pylint disable is not that much more work then),

But I get that this might be a little opinionated and that we'd get complaints from this, thank you for explaining.

> Since we're talking about addressing false negatives, not false positives, that seems like a smart way of iterating to me.

:+1:

Reply:
This discussion helped clarify for me that beefing up the checker to assume try statements always fail is worthy of discussion, but would need to be solved in tandem with #5524, so I suggest we keep discussing there for 2.14 or later.